27 migrate the existing codebase to orm prism

radu-constantin opened the pull request:
Setting up two different environments for testing and development was a bit tricky. Prisma can only load environment variables directly from process.env. It cannot load from a config file like we attempted to do. The solution for this was to declare two .env files. One for the dev environment and one for the test environment. I tried my best to find another solution, but I couldn’t find anything.
https://www.prisma.io/docs/guides/testing/integration-testing
https://www.prisma.io/docs/guides/development-environment/environment-variables/using-multiple-env-files
Most of the work was done in the package.json `test` script. 
"dotenv -e ./.env.test -- npx prisma migrate deploy && dotenv -e ./.env.test jest --verbose --runInBand"
The script itself does 2 things: 
1.	dotenv -e ./.env.test -- npx prisma migrate deploy
Loads the environment variables from .env.test into process.env. (this is necessary in order to get the test db URI necessary for configuring the test db and running the tests). We also run `prisma migrate deploy` in order to load the Prisma schema into the database server on docker.
2.	dotenv -e ./.env.test jest --verbose --runInBand"
Loads the environment variables from .env.test again and runs the tests.
The dev script now also runs `npx prisma migrate deploy` to make sure the prisma schema is loaded in the server. It is not necessary to be run every time, but I think its safer that way. 
In order to visually check out the entire database you can use the command `npx prisma studio` and a GUI will open in your browser showcasing the db. By default this command will showcase the development database. To check the test database I’ve added another script to package.json `prisma-studio-test`. This will let you analyze the test db if you ever need to. (I don’t think it’s necessary to be honest, but I used it to debug some issues when setting everything up).

For the test I decided to create a separate seed file, since we will probably want to use different data than for development. It is found in the tests folder. 
At the moment, the test file seeds the test database before all tests are run in order to have some initial data.
Also, in the docker-compose file I’ve deleted the link with the initial SQL database since we don’t want that one to be pre-loaded in the container anymore.

Make sure to run `npm install` and delete the current container instance and associated volumes so you can start fresh.

nina-dk commented on prisma/seed.ts:
I'm getting a complaint from ESLint here "'@prisma/client' should be listed in the project's dependencies. Run 'npm i -S @prisma/client' to add it". Should we do that?

radu-constantin replied:
That is very strange. Looking at my package.json file, I already have it included in the dependencies list. We need to investigate this.

MaxHass12 replied:
This is happening with me as well. This also happened with me initially while setting up the project, when @radu-constantin had to manually add some dependencies ( related to lint )to `package.json`.

nina-dk replied:
For some reason, I'm not seeing it anymore.

nina-dk commented on prisma/seed.ts:
```diff
@@ -13,9 +13,11 @@ async function main() {
   await prisma.gender.createMany({
     data: [
       {
+        id: 1,
```
Do we need to add the `id`s manually?

radu-constantin replied:
Not really, I did it out of habit :). WIll fix!

nina-dk commented on prisma/schema.prisma:
```diff
@@ -16,71 +16,69 @@ model Gender {
-  location            String?               @db.VarChar(50)
-  users               NeighborhoodUsers[]
+  id                  Int                 @id @default(autoincrement())
+  admin_id            Decimal             @db.Decimal
```
why is that Decimal instead of Int?

radu-constantin replied:
In the initial SQL schema, the id was defined as being numeric, not an integer. Prisma converted numeric to decimal. I think we can safely change it to integer.

radu-constantin commented:
The test database is seeded automatically when you run the test. You don't have to do it manually. Maybe that's the issue.
Delete your volume and container and try again please.
The dev environment needs to be seeded manually though. I could of included seeding as part of the `npm run dev` script, but since seeding will basically reset all data in the db, I would prefer that we do it manually.
This adds a bit of complexity when starting the dev server for the first time. Hmmm, I will make two separate scripts, `npm run dev` and `npm run dev seed`. The second one will also seed the database, while the first will just start the server, without touching the data.

MaxHass12 commented:
> The test database is seeded automatically when you run the test. You don't have to do it manually. Maybe that's the issue.
> Delete your volume and container and try again please.
> The dev environment needs to be seeded manually though. I could of included seeding as part of the `npm run dev` script, but since seeding will basically reset all data in the db, I would prefer that we do it manually.
> This adds a bit of complexity when starting the dev server for the first time. Hmmm, I will make two separate scripts, `npm run dev` and `npm run dev seed`. The second one will also seed the database, while the first will just start the server, without touching the data.

Done.

MaxHass12 commented:
In `src/model/db_query.ts` at `line 15`, `config.POSTGRES_URL` should be `config.DATABASE_URL` because we changed the name of the exported variable from the module `src/utils/config.ts`.

All rest looks good. I am not in a position to comment on rest of code because I am not familiar enough with ORM Prisma at the moment. I will pick ORM up in a couple of days after I am done going through the TS docs.

nina-dk commented:
> The test database is seeded automatically when you run the test. You don't have to do it manually. Maybe that's the issue. Delete your volume and container and try again please. The dev environment needs to be seeded manually though. I could of included seeding as part of the `npm run dev` script, but since seeding will basically reset all data in the db, I would prefer that we do it manually. This adds a bit of complexity when starting the dev server for the first time. Hmmm, I will make two separate scripts, `npm run dev` and `npm run dev seed`. The second one will also seed the database, while the first will just start the server, without touching the data.

Could you comment above each script that you're adding so that we know (and remember) what it does? It's becoming hard to remember what to use when with the increasing number of scripts.

radu-constantin commented:
I've added the the extra scripts and I've updated the README.md. I can't directly write comments in package.json (json files do not allow comments).
Shwetank, at this point I don't think we need either the `db_query` module nor `neighborhood` ts. Using Prisma leaves both modules obsolete. I'm not 100% sure about this though, that's why I would like to leave them there for now.

nina-dk commented on package.json:
I'm getting an error when running `npm run dev-seed`:
```
→ 66 const bobNeighborhood = await prisma.neighborhood.create(
Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(Error { kind: Db, cause: Some(DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState(E22P03), message: "incorrect binary data format in bind parameter 1", detail: None, hint: None, position: None, where_: Some("unnamed portal parameter $1"), schema: None, table: None, column: None, datatype: None, constraint: None, file: Some("postgres.c"), line: Some(1912), routine: Some("exec_bind_message") }) }), transient: false })
```
I connected to the dev-db afterwards to see what it looks like. The users have been created as have the genders, but no neighbourhoods exist. I run `npm run start-db` before all else.

radu-constantin replied:
Could be, can you delete the volume and try again?

nina-dk reviewed (commented):
Could the change from Decimal to Int be causing this error I mentioned?

radu-constantin commented:
I removed the `npm run dev-seed` script and replaced it with a normal `seed` script. It didn't make sense to have a script that seeds and also starts the dev server.
